### PR TITLE
Implement new sign up page

### PR DESCRIPTION
**Pull request by weeman1337**
closes https://github.com/engelsystem/engelsystem/issues/1124

[pipeline](https://chaos.expert/engelsystem/engelsystem/-/pipelines/40112)

### CODE REVIEW

File: resources/lang/de_DE/default.po [hunk 1]
```diff
@@ -2186,6 +2164,9 @@ msgstr "Passwort"
 msgid "settings.password.info"
 msgstr "Hier kannst Du Dein Passwort ändern."
 
+msgid "settings.password.confirmation-does-not-match"
+msgstr "Passwort und Passwordwiederholung stimmen nicht überein."
```
Comment (MyIgel) [inline resources/lang/de_DE/default.po:2168]: `Passwortwiederholung`

File: resources/views/macros/form.twig [hunk 2]
```diff
@@ -214,9 +214,10 @@ Renders a Bootstrap checkbox element with mb-3.
 @param {bool} [opt.disabled=false] - Whether to add the "disabled" attribute. Defaults to false.
 @param {bool} [opt.raw_label=false] - Whether to use the raw label value (=do not escape). Defaults to false.
 @param {string} [opt.info] - If set an additional info icon will be added to the label with the text as tooltip.
+@param {string} [opt.class="mb-3"] - CSS classes for the checkbox element. Defaults to "mb-3".
 #}
 {% macro checkbox(name, label, opt) %}
-    <div class="form-check mb-3">
+<div class="form-check {{ opt.class is defined ? opt.class : 'mb-3' }}">
```
Comment (MyIgel) [inline resources/views/macros/form.twig:220]: Spacing

**Review by MyIgel — CHANGES_REQUESTED** [review]
Some first impressions / preliminary tests:
* [x] The email preferences are not readable on the dark theme
* [x] To be consistent with the rest of the page, all required fields should be marked with an `*`
* [x] Currently the horizontal distance between the angeltypes makes them somewhat disconnected which is confusing. Instead of splitting ot two they sould be splitted into three  cols(and the last left blank for desktop size)
* [x] "Show mobile number" is not saved correctly on registration
* [x] Pronoun is shown, even if disabled
* [ ] The headline about private messages is shown above the arrival date on config
  ```
      'enable_dect' => false,
    'enable_mobile_show' => false,
    'enable_user_name' => false,
    'enable_pronoun' => false,
    'enable_planned_arrival' => true,
  ```
  * [x] `The email "%s" is already taken.` benötigt npoch einen ersetzten platzhalter (oder entfernen des platzhalters) & ist nicht besonders aussagekräftig als fehlermeldung für "der nickname wird bereits verwendet)

File: resources/views/macros/form.twig (hunk 2 above)
Comment (weeman1337) [reply resources/views/macros/form.twig:220]: We need  a twig formatter / linter :see_no_evil:

Comment (weeman1337) [reply resources/views/macros/form.twig:220]: Done

File: resources/lang/de_DE/default.po (hunk 1 above)
Comment (weeman1337) [reply resources/lang/de_DE/default.po:2168]: Done

**Review by msquare — CHANGES_REQUESTED** [review]
- [x] When I enter anything invalid, like missing mail address, the selected angeltypes are reset, so I need to check everything again. I would expect that the page remembers the already selected angeltypes

**Comment by weeman1337** [comment]
> * [ ]  When I enter anything invalid, like missing mail address, the selected angeltypes are reset, so I need to check everything again. I would expect that the page remembers the already selected angeltypes

It didn't clear the session variables, that hold the form data after checking the boxes after an error. Should work now™

**Review by msquare — APPROVED** [review]

**Review by MyIgel — CHANGES_REQUESTED** [review]
That is a more thorough review, looks great so far!

Was there a specific reason why you changed the path from `/register` to `/sign.up`?

Review todo: SSO magic & "feature" coverage (edge cases / checks / configurable options)

File: src/Factories/User.php [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+        private SessionInterface $session,
+        private Validator $validator,
+    ) {
+    }
+
+    /**
+     * Takes some arbitrary data, validates it and tries to create a user from it.
+     *
+     * @param Array<string, mixed> $rawData Raw data from which a user should be created
+     * @return The user if successful
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply src/Factories/User.php:42]: This would be duplicate information since the signature already says that it returns an `EngelsystemUser`.

**Comment by msquare** [comment]
- [ ] OAuth test was successful.

File: src/Factories/User.php [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+        if (!is_null($oAuthEnablePassword)) {
+            // o-auth overwrites config
+            $isPasswordEnabled = $oAuthEnablePassword;
+        }
+
+        return $isPasswordEnabled;
+    }
+
+    public function determineBuildUpStartDate(): DateTimeInterface
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply src/Factories/User.php:65]: I would create a follow-up issue to extract this if you are okay with it. Thumb up here :slightly_smiling_face:

File: src/Factories/User.php [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+            'email_by_human_allowed' => 'optional|checked',
+            'email_messages' => 'optional|checked',
+            'email_news' => 'optional|checked',
+            'email_goody' => 'optional|checked',
+            // Using length here, because min/max would validate dect/mobile as numbers.
+            'mobile' => 'optional|length:0:40',
+        ];
+
+        $isPasswordEnabled = $this->determineIsPasswordEnabled();
+
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply src/Factories/User.php:90]: Imho assignments followed by a control structure are hard to parse. But I don't are and if you want no newline between it I can remove it :shrug:

File: src/Factories/User.php [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+        if ($isShowMobileEnabled) {
+            $validationRules['mobile_show'] = 'optional|checked';
+        }
+
+        $isPlannedArrivalDateEnabled = $this->config->get('enable_planned_arrival');
+
+        if ($isPlannedArrivalDateEnabled) {
+            $isoBuildUpStartDate = $this->determineBuildUpStartDate();
+            $tearDownEndDate = $this->config->get('teardown_end');
+            /** @var DateTimeInterface|null $tearDownEndDate */
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply src/Factories/User.php:121]: Done

File: src/Factories/User.php [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+            if ($isoBuildUpStartDate && $tearDownEndDate) {
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply src/Factories/User.php:123]: Done

File: src/Factories/User.php [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+        $settings = new Settings([
+            'language'        => $this->session->get('locale') ?? 'en_US',
+            'theme'           => $this->config->get('theme'),
+            'email_human'     => $data['email_by_human_allowed'] ?? false,
+            'email_messages'  => $data['email_messages'] ?? false,
+            'email_goody'     => $data['email_goody'] ?? false,
+            'email_shiftinfo' => $data['email_shiftinfo'] ?? false,
+            'email_news'      => $data['email_news'] ?? false,
+            'mobile_show'     => $isShowMobileEnabled && !!($data['mobile_show'] ?? false),
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply src/Factories/User.php:255]: Done

**Comment by weeman1337** [comment]
> Was there a specific reason why you changed the path from `/register` to `/sign.up`?

We discussed this and I did it. Maybe it is risky to change it during an event. I could add a redirect from `/register` to `/sign-up`.

> Review todo: SSO magic & "feature" coverage (edge cases / checks / configurable options)

@msquare and I fixed and tested the oauth stuff this morning.

File: src/Controllers/SignUpController.php [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+        // form-data-register-submit is a marker, that the form was submitted.
+        // It will be used for instance to use the default angel types or the user selected ones.
+        // Clear it before render to reset the marker state.
+        $this->session->remove('form-data-register-submit');
+
+        return $this->response->withView(
+            'pages/sign-up',
+            [
+                'tShirtSizes' => array_merge(['' => __('Please select...')], $this->config->get('tshirt_sizes')),
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply src/Controllers/SignUpController.php:73]: Done

File: resources/views/pages/sign-up.twig [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+                                'value': f.formData('lastname', ''),
+                            }
+                        ) }}
+                    </div>
+                </div>
+            </div>
+        {% endif %}
+
+        <div class="mb-5">
+            <h2>{{ __('Event data') }}</h2>
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply resources/views/pages/sign-up.twig:179]: Done

File: src/Factories/User.php [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+
+        $settings = new Settings([
+            'language'        => $this->session->get('locale') ?? 'en_US',
+            'theme'           => $this->config->get('theme'),
+            'email_human'     => $data['email_by_human_allowed'] ?? false,
+            'email_messages'  => $data['email_messages'] ?? false,
+            'email_goody'     => $data['email_goody'] ?? false,
+            'email_shiftinfo' => $data['email_shiftinfo'] ?? false,
+            'email_news'      => $data['email_news'] ?? false,
+            'mobile_show'     => $isShowMobileEnabled && (bool) $data['mobile_show'],
```
Comment (MyIgel) [inline src/Factories/User.php:254]:
```suggestion
            'mobile_show'     => $isShowMobileEnabled && $data['mobile_show'],
```

File: includes/pages/guest_login.php [hunk 12]
```diff
@@ -1,543 +0,0 @@
...
-            success(__('Angel registration successful!'));
-
-            // User is already logged in - that means a supporter has registered an angel. Return to register page.
-            if ($authUser) {
-                throw_redirect(page_link_to('register'));
-            }
-
-            // If a welcome message is present, display it on the next page
-            if ($config->get('welcome_msg')) {
-                $session->set('show_welcome', true);
```
Comment (MyIgel) [inline includes/pages/guest_login.php:345]: That functionality is missing

File: includes/pages/guest_login.php [hunk 13]
```diff
@@ -1,543 +0,0 @@
...
-
-            engelsystem_log(
-                'User ' . User_Nick_render($user, true)
-                . ' signed up as: ' . join(', ', $user_angel_types_info)
-            );
-            success(__('Angel registration successful!'));
-
-            // User is already logged in - that means a supporter has registered an angel. Return to register page.
-            if ($authUser) {
-                throw_redirect(page_link_to('register'));
```
Comment (MyIgel) [inline includes/pages/guest_login.php:340]: This seems to be missing in the new version but is a better default in that case

File: includes/pages/guest_login.php [hunk 14]
```diff
@@ -1,543 +0,0 @@
...
-            }
-
-            // If a welcome message is present, display it on the next page
-            if ($config->get('welcome_msg')) {
-                $session->set('show_welcome', true);
-            }
-
-            // Login the user
-            if ($user->oauth->count()) {
```
Comment (MyIgel) [inline includes/pages/guest_login.php:349]: That is not covered too

**Review by MyIgel — CHANGES_REQUESTED** [review]

**Comment by MyIgel** [comment]
(& there are merge conflicts @weeman1337 ;))

File: includes/pages/guest_login.php (hunk 13 above)
Comment (weeman1337) [reply includes/pages/guest_login.php:340]: Yes, could be a bit strange to land on the sign-up page again.

Comment (MyIgel) [reply includes/pages/guest_login.php:340]: What part of it would be anoying? Its imho more anoying if i get to the / page as it has even less value (as in_: what would i want do do from there?)

Comment (weeman1337) [reply includes/pages/guest_login.php:340]: „Dies scheint in der neuen Version zu fehlen, ist aber in diesem Fall ein besserer Standard“

Reads like the new one is better… So it should be the same behaviour as in guest login?

Comment (weeman1337) [reply includes/pages/guest_login.php:340]: r-ledigt

File: src/Factories/User.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (MyIgel) [reply src/Factories/User.php:89]: A comment would (in that case at least) would not make anything clearer?

**Comment by MyIgel** [comment]
This PR finally looks nice but the `/users` page has a link to `/register ` which should be patched ^^

**Review by MyIgel — CHANGES_REQUESTED** [review]
Just the users page needs a patch

**Comment by xuwhite** [comment]
I dont like the renaming from signup to register, because for me signup is more a first time thing and this page is used also from shicos that do it for other users
